spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java

```diff
@@ -524,18 +526,111 @@ private RedisLockRegistry getOuterType() {
+			super(path);
+		}
+
+		@Override
+		protected boolean tryRedisLockInner(long time) throws ExecutionException, InterruptedException {
+			return subscribeLock(time);
+		}
+
+		@Override
+		protected void removeLockKeyInnerUnlink() {
+			RedisLockRegistry.this.redisTemplate.execute(
+					unLinkUnLockScript, Collections.singletonList(this.lockKey),
+					RedisLockRegistry.this.unLockChannelKey);
+		}
+
+		@Override
+		protected void removeLockKeyInnerDelete() {
+			RedisLockRegistry.this.redisTemplate.execute(
+					deleteUnLockScript, Collections.singletonList(this.lockKey),
+					RedisLockRegistry.this.unLockChannelKey);
+
+		}
+
+		private boolean subscribeLock(long time) throws ExecutionException, InterruptedException {
+			final long expiredTime = System.currentTimeMillis() + time;
+			if (obtainLock()) {
+				return true;
+			}
+
+			if (!(RedisLockRegistry.this.isRunningRedisMessageListenerContainer
+				&& RedisLockRegistry.this.redisMessageListenerContainer != null
+				&& RedisLockRegistry.this.redisMessageListenerContainer.isRunning())) {
+				runRedisMessageListenerContainer();
+			}
+			while (time == -1 || expiredTime >= System.currentTimeMillis()) {
+				try {
+					Future<String> future =
+							RedisLockRegistry.this.unlockNotifyMessageListener.subscribeLock(this.lockKey);
+					//DCL
+					if (obtainLock()) {
+						return true;
+					}
+					try {
+						//if short expireAfter key expire for ttl, no receive unlock msg
+						long waitTime = time >= 0 ? time : RedisLockRegistry.this.expireAfter;
+						future.get(waitTime, TimeUnit.MILLISECONDS);
+					}
+					catch (TimeoutException ignore) {
+					}
+					if (obtainLock()) {
+						return true;
+					}
+				}
+				finally {
+					RedisLockRegistry.this.unlockNotifyMessageListener.unSubscribeLock(this.lockKey);
+				}
+			}
+			return false;
+		}
+
 		private void runRedisMessageListenerContainer() {
-			synchronized (RedisLockRegistry.this.redisMessageListenerContainer) {
+			synchronized (RedisLockRegistry.this.locks) {
 				if (!(RedisLockRegistry.this.isRunningRedisMessageListenerContainer
-						&& RedisLockRegistry.this.redisMessageListenerContainer.isRunning())) {
-					RedisLockRegistry.this.redisMessageListenerContainer.afterPropertiesSet();
+					&& RedisLockRegistry.this.redisMessageListenerContainer != null
+					&& RedisLockRegistry.this.redisMessageListenerContainer.isRunning())) {
+
+					if (RedisLockRegistry.this.redisMessageListenerContainer == null) {
+						setupUnlockMessageListener(RedisLockRegistry.this.redisTemplate.getConnectionFactory());
+						RedisLockRegistry.this.redisMessageListenerContainer.afterPropertiesSet();
+					}
+
 					RedisLockRegistry.this.redisMessageListenerContainer.start();
 					RedisLockRegistry.this.isRunningRedisMessageListenerContainer = true;
 				}
 			}
 		}
+
 	}
 
+	/**
+	 * Used by RedisPubSubLock
+	 */
 	private static final class RedisUnLockNotifyMessageListener implements MessageListener {
```

Comment:
This is `private` class - cannot it move into `RedisPubSubLock` class directly to preserve a context?